src/PBTRandom.sol

```diff
@@ -0,0 +1,271 @@
+    //        - if the value is zero, it means randIndex has not been used, so we can return randIndex
+    //        - if the value is non-zero, it means the value has not been used, so we can return _availableRemainingTokens[randIndex]
+    //    - update the _availableRemainingTokens mapping state
+    //        - set _availableRemainingTokens[randIndex] to either the index or the value of the last entry in the mapping (depends on the last entry's state)
+    //        - decrement _numAvailableRemainingTokens to mimic the shrinking of an array
+    function _useRandomAvailableTokenId() internal returns (uint128) {
+        uint128 numAvailableRemainingTokens = _numAvailableRemainingTokens;
+
+        // Devs can swap this out for something less gameable like chainlink if it makes sense for their use case.
+        uint256 randomNum = uint256(
+            keccak256(
+                abi.encode(
+                    _msgSender(),
+                    tx.gasprice,
+                    block.number,
+                    block.timestamp,
+                    block.difficulty,
+                    blockhash(block.number - 1),
+                    address(this),
+                    numAvailableRemainingTokens
+                )
+            )
+        );
+        uint128 randomIndex = uint128(randomNum % numAvailableRemainingTokens);
+        return _getAvailableTokenAtIndex(randomIndex);
+    }
+
+    function _getAvailableTokenAtIndex(uint128 indexToUse) internal returns (uint128) {
+        uint128 numAvailableRemainingTokens = _numAvailableRemainingTokens;
+        if (indexToUse >= numAvailableRemainingTokens) {
+            revert InvalidRandomIndex();
+        }
+
+        uint128 valAtIndex = _availableRemainingTokens[indexToUse];
+        uint128 result;
+        if (valAtIndex == 0) {
+            // This means the index itself is still an available token
+            result = indexToUse;
+        } else {
+            // This means the index itself is not an available token, but the val at that index is.
+            result = valAtIndex;
+        }
+
+        uint128 lastIndex = numAvailableRemainingTokens - 1;
+        if (indexToUse != lastIndex) {
+            // Replace the value at indexToUse, now that it's been used.
+            // Replace it with the data from the last index in the array, since we are going to decrease the array size afterwards.
+            uint128 lastValInArray = _availableRemainingTokens[lastIndex];
+            if (lastValInArray == 0) {
+                // This means the index itself is still an available token
+                _availableRemainingTokens[indexToUse] = lastIndex;
+            } else {
+                // This means the index itself is not an available token, but the val at that index is.
+                _availableRemainingTokens[indexToUse] = lastValInArray;
+                delete _availableRemainingTokens[lastIndex];
+            }
+        }
+
+        _numAvailableRemainingTokens--;
```

Comment:
Moved this operation from `_useRandomAvailableTokenId` to this function to group all state changes in one place.

Reply:
Agree with other comment to merge these two methods

If not, let's at a minimum rename  `_getAvailableTokenAtIndex` to `_useAvailableTokenAtIndex` since it changes state